substrate-node/runtime/src/migrations/tfgrid_v15_smart_contract_v8.rs

```diff
@@ -0,0 +1,240 @@
+    info!(" <<< Twin migration success, storage version upgraded");
+    return (0, twins.len() as u64 + 1);
+}
+
+pub fn migrate_tfgrid<T: Config>(
+    twins: &mut BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>>,
+) -> (u64, u64) {
+    info!(" >>> Migrating twin storage...");
+
+    let mut reads = 0;
+    for (_twin_id, twin) in Twins::<T>::iter() {
+        debug!("migrating twin: {:?}", twin);
+
+        let new_twin = pallet_tfgrid::types::Twin::<AccountIdOf<T>> {
+            id: twin.id,
+            account_id: twin.account_id,
+            relay: None,
+            entities: twin.entities,
+            pk: None,
+        };
+        twins.insert(twin.id, new_twin);
+        reads += 1;
+    }
+
+    // Return the weight consumed by the migration.
+    return (reads, 0);
+}
+
+pub fn migrate_smart_contract<T: Config>(
+    twins: &BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>>,
+) -> (u64, u64) {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        pallet_smart_contract::PalletVersion::<T>::get()
+    );
+
+    let mut reads = 0;
+    let mut writes = 0;
+    let mut twin_contract_locked_balances: BTreeMap<u32, BalanceOf<T>> = BTreeMap::new();
+    // Fetch all locked balances based on the contract locks in storage and accumulate them by twin id
+    for (ctr_id, l) in pallet_smart_contract::ContractLock::<T>::iter() {
+        let ctr = pallet_smart_contract::Contracts::<T>::get(ctr_id);
+        reads += 1;
+        match ctr {
+            Some(contract) => {
+                if !twins.contains_key(&contract.twin_id) {
+                    debug!(
+                        "twins: {} does not exist, removing contract and lock",
+                        contract.twin_id
+                    );
+                    pallet_smart_contract::Contracts::<T>::remove(ctr_id);
+                    pallet_smart_contract::ContractLock::<T>::remove(ctr_id);
+                    writes += 2;
+                } else {
+                    *twin_contract_locked_balances
+                        .entry(contract.twin_id)
+                        .or_default() += l.amount_locked;
+                }
+            }
+            None => {
+                debug!(
+                    "no contract found for contract lock {}, cleaning up lock..",
+                    ctr_id
+                );
+                writes += 1;
+                pallet_smart_contract::ContractLock::<T>::remove(ctr_id);
+            }
+        }
+    }
+
+    for (twin_id, twin) in twins {
+        // If the twin needs to have some locked balance on his account because of running contracts
+        // Check how much we can actually lock based on his current total balance
+        // this will make sure the locked balance will not exceed the total balance on the twin's account
+        let should_lock = twin_contract_locked_balances.get(&twin_id).map(|b| {
```

Comment:
Maybe be we should not consider the twins without locked balance before doing this?
I mean in case of `twin_contract_locked_balances.get(&twin_id) == None`
Not sure if all twins have contracts and locked balances...

Reply:
We fill the twin_contract_locked_balances map before this step and we only add an element if the twin exists. So the None case will not happen.

Reply:
Yes, I get this
But in the loop `for (twin_id, twin) in twins { ...` all the twins are considered right?

Reply:
Was expecting something like `for twin_id in twin_contract_locked_balances.keys() { ...` instead
But maybe I miss something

Reply:
Yes but we do a get, if the get doesn't find the key it will return None directly and then we don't do anything with it (see the if let some(should_lock)). The map part after the get is only executed if the get returned something.

Reply:
> Was expecting something like `for twin_id in twin_contract_locked_balances.keys() { ...` instead But maybe I miss something

Ah like that. I don't know if this part should happen for all twins or only for the twins that have a contract. @DylanVerstraete?

Reply:
Great, I see now
Thanks for the explanation of the `return None directly` part ;)